package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -0,0 +1,943 @@
+            self.distances_buffer.append(np.sqrt(dists_buffer[elm])[sorted_indices])
+
+    def get_indices(self):
+        if self.indices_buffer is None:
+            self.create_buffers()
+        return self.indices_buffer
+
+    def get_distances(self):
+        if self.distances_buffer is None:
+            self.create_buffers()
+        return self.distances_buffer
+
+    def get_coordinates(self):
+        if self.coordinates_buffer is None:
+            self.create_buffers()
+        return self.coordinates_buffer
+
+
+
+cdef class NSGrid(object):
+    cdef bint debug
+    cdef readonly real cutoff
+    cdef ns_int size
+    cdef ns_int ncoords
+    cdef ns_int[DIM] ncells
+    cdef ns_int[DIM] cell_offsets
+    cdef real[DIM] cellsize
+    cdef ns_int nbeads_per_cell
+    cdef ns_int *nbeads # size
+    cdef ns_int *beadids # size * nbeads_per_cell
+    cdef ns_int *cellids # ncoords
+
+    def __init__(self, ncoords, cutoff, PBCBox box, max_size, debug=False):
+        cdef ns_int i, x, y, z
+        cdef ns_int ncellx, ncelly, ncellz, size, nbeadspercell
+        cdef real bbox_vol
+        self.debug = debug
+
+        self.ncoords = ncoords
+
+        # Calculate best cutoff
+        self.cutoff = cutoff
+        bbox_vol = box.c_pbcbox.box[XX][XX] * box.c_pbcbox.box[YY][YY] * box.c_pbcbox.box[YY][YY]
+        size = bbox_vol/cutoff**3
+        nbeadspercell = ncoords/size
+        while bbox_vol/self.cutoff**3 > max_size:
+            self.cutoff *= 1.2
+
+        for i in range(DIM):
+            self.ncells[i] = <ns_int> (box.c_pbcbox.box[i][i] / self.cutoff)
+            if self.ncells[i] == 0:
+                self.ncells[i] = 1
+            self.cellsize[i] = box.c_pbcbox.box[i][i] / self.ncells[i]
+        self.size = self.ncells[XX] * self.ncells[YY] * self.ncells[ZZ]
+
+        if self.debug:
+            print("NSGrid: Requested cutoff: {:.3f} (Ncells={}, Avg # of beads per cell={}), Optimized cutoff= {:.3f} (Ncells={}, Avg # of beads per cell={})".format(
+                cutoff, size, nbeadspercell,
+                self.cutoff, self.size, <ns_int> (ncoords / self.size)
+            ))
+            print("NSGrid: Size={}x{}x{}={}".format(self.ncells[XX], self.ncells[YY], self.ncells[ZZ], self.size))
+
+        self.cell_offsets[XX] = 0
+        self.cell_offsets[YY] = self.ncells[XX]
+        self.cell_offsets[ZZ] = self.ncells[XX] * self.ncells[YY]
+
+        # Allocate memory
+        self.nbeads = <ns_int *> PyMem_Malloc(sizeof(ns_int) * self.size)
+        if not self.nbeads:
+            raise MemoryError("Could not allocate memory from NSGrid.nbeads ({} bits requested)".format(sizeof(ns_int) * self.size))
+        self.beadids = NULL
+        self.cellids = <ns_int *> PyMem_Malloc(sizeof(ns_int) * self.ncoords)
+        if not self.cellids:
+            raise MemoryError("Could not allocate memory from NSGrid.cellids ({} bits requested)".format(sizeof(ns_int) * self.ncoords))

```

Comment:
another memory leak. If allocating previous arrays works their memory will not be released again in the python process. I assume there are more memory leaks like this in the code.

Reply:
Is this also true when the call to `__init__` fails? I believe you for other functions but for `__init__` I do not know what the interpreter will do. The question is also if some of the allocations should go into a `__cinit__` function.

Reply:
I think that __dealloc__ is always called, so even when __init__ failed... but I will check that

Reply:
According to an old [mailing list entry](https://groups.google.com/forum/#!searchin/cython-users/__init__|sort:date/cython-users/VzwYgi8E6fY/wtX8BeNADAAJ) this dealloc should be called and work fine. So no memory leak. But according to the [docs](http://docs.cython.org/en/latest/src/userguide/special_methods.html#initialisation-methods-cinit-and-init) the `__init__` method can be called twice. 

> Under some circumstances it is possible for __init__() to be called more than once or not to be called at all, so your other methods should be designed to be robust in such situations.

If we call malloc twice that would be a memory leak. More troubling is that the memory might not be initialized at all.